### PR TITLE
feat(farmingcalculator): add dev mining estimation calculation for rewards

### PR DESCRIPTION
**Pull request by daywiss**
**Motivation** 
#196 

**Changes**
- adds a developer mining estimation calculator (based on 50k UMA, split between EMPs, divided by half for LM rewards)
- adds hardcoded whitelist contracts. currently only shows for our DEC contracts
- shoehorns results into existing liquidity mining calculation
- adds tooltips and links user to more information links to [https://twitter.com/UMAprotocol/status/1324199135080148993](https://twitter.com/UMAprotocol/status/1324199135080148993)

**Screens**
image
image

### CODE REVIEW

File: features/yield/FarmingCalculator.tsx [hunk 1]
```diff
@@ -225,6 +213,9 @@ const FarmingCalculator = () => {
           // Reward has not started.
           continue;
         } else {
+          if (reward.token == "UMA") {
+            reward.count = liquidityRewards;
```
Comment (daywiss) [inline features/yield/FarmingCalculator.tsx:217]: just updating the hardcoded reward count with our dynamically calculated liquidity rewards from devmining. This is what gets displayed in the UI

File: features/yield/FarmingCalculator.tsx [hunk 2]
```diff
@@ -314,7 +311,207 @@ const FarmingCalculator = () => {
       </Box>
     );
   } else {
-    return <LiquidityMiningV2 />;
+    return (
```
Comment (daywiss) [inline features/yield/FarmingCalculator.tsx:314]: adds back in previous liquidity mining code

File: utils/calculators.test.ts [hunk 3]
```diff
@@ -40,3 +54,47 @@ test("apr", () => {
...
 });
+
+test("getCoinGeckoTokenPrice", async () => {
+  const contracts = [
+    "0xEB4C2781e4ebA804CE9a9803C67d0893436bB27D", // rBTC
+    "0xC02aaA39b223FE8D0A0e5C4F27eAD9083C756Cc2", // wETH
+    "0xeca82185adCE47f39c684352B0439f030f860318", // PERL
+  ];
+  for (let contract of contracts) {
+    const result = await getSimplePriceByContract(contract);
```
Comment (daywiss) [inline utils/calculators.test.ts:65]: add a new coingecko call to get instant price from contract address

File: utils/calculators.ts [hunk 4]
```diff
@@ -31,3 +32,91 @@ export const calcApr = (
...
+// }
+type DevMiningCalculatorParams = {
+  ethers: any;
+  getPrice: any;
+  empAbi: any;
+  erc20Abi: any;
+  provider: any;
+};
+export function DevMiningCalculator({
+  provider,
```
Comment (daywiss) [inline utils/calculators.ts:46]: This is most of the dev mining estimate calculation. Outputs all emps and their estimated rewards.

File: containers/DevMining.ts [hunk 5]
```diff
@@ -0,0 +1,56 @@
...
+const useDevMiningCalculator = ({
+  totalRewards = defaultTotalRewards,
+  empWhitelist = defaultEmpWhitelist,
+} = {}) => {
+  const { provider } = Connection.useContainer();
+  const [devMiningRewards, setRewards] = useState<Map<string, string> | null>();
+  const [devMiningCalculator, setCalculator] = useState<any | null>();
+
+  useEffect(() => {
+    if (provider == null) return;
```
Comment (nicholaspai) [inline containers/DevMining.ts:29]:
```suggestion
    if (provider === null) return;
```

File: features/yield/FarmingCalculator.tsx [hunk 6]
```diff
@@ -206,9 +187,16 @@ const FarmingCalculator = () => {
     if (address) {
       setTokenAddress(address.toLowerCase());
     }
-    if (Object.keys(WEEKLY_UMA_REWARDS).includes(tokenAddress)) {
+    if (devMiningRewards == null || empAddress == null) return;
```
Comment (nicholaspai) [inline features/yield/FarmingCalculator.tsx:190]:
```suggestion
    if (devMiningRewards === null || empAddress === null) return;
```

File: features/yield/FarmingCalculator.tsx [hunk 7]
```diff
@@ -314,7 +311,207 @@ const FarmingCalculator = () => {
...
+            here
+          </a>
+          .
+        </Typography>
+        {timeUntilRoll && timeUntilRoll > 0 && (
+          <>
+            <br></br>
+            <br></br>
+            <Typography>
+              <strong>Liquidity mining rewards during the roll:</strong> Before{" "}
```
Comment (nicholaspai) [inline features/yield/FarmingCalculator.tsx:339]: Do you think its worth it adding any details about dev mining here for the user to better understand their rewards?

**Review by nicholaspai — APPROVED** [review]
Demo site looks great!

File: containers/DevMining.ts (hunk 5 above)
Comment (daywiss) [reply containers/DevMining.ts:29]: Strict equality on null makes me nervous since i want undefined to be handled the same case as null

Comment (nicholaspai) [reply containers/DevMining.ts:29]: `provider` is type `Provider` or `null` though, and more importantly its consistent with the rest of the repo with === vs ==

File: features/yield/FarmingCalculator.tsx [hunk 8]
```diff
@@ -225,6 +213,9 @@ const FarmingCalculator = () => {
           // Reward has not started.
           continue;
         } else {
+          if (reward.token == "UMA") {
```
Comment (nicholaspai) [inline features/yield/FarmingCalculator.tsx:216]:
```suggestion
          if (reward.token === "UMA") {
```

**Pull request merged**